test/e2e/controller/etcd_test.go

```diff
@@ -0,0 +1,689 @@
+
+	exitCode := m.Run()
+
+	cancelAnnotator()
+
+	testEnv.Close()
+	os.Exit(exitCode)
+}
+
+// TestBasic tests creation, hibernation, unhibernation and deletion of the Etcd cluster.
+func TestBasic(t *testing.T) {
+	t.Parallel()
+	log := testr.NewWithOptions(t, testr.Options{LogTimestamp: true})
+
+	testCases := []struct {
+		name       string
+		replicas   int32
+		tlsEnabled bool
+	}{
+		{
+			name:       "no-tls-0",
+			tlsEnabled: false,
+			replicas:   0,
+		},
+		{
+			name:       "no-tls-1",
+			tlsEnabled: false,
+			replicas:   1,
+		},
+		{
+			name:       "no-tls-3",
+			tlsEnabled: false,
+			replicas:   3,
+		},
+		{
+			name:       "tls-0",
+			tlsEnabled: true,
+			replicas:   0,
+		},
+		{
+			name:       "tls-1",
+			tlsEnabled: true,
+			replicas:   1,
+		},
+		{
+			name:       "tls-3",
+			tlsEnabled: true,
+			replicas:   3,
+		},
+	}
+
+	g := NewWithT(t)
+
+	for _, provider := range providers {
+		for _, tc := range testCases {
+			tcName := fmt.Sprintf("basic-%s-%s", tc.name, provider)
+			t.Run(tcName, func(t *testing.T) {
+				t.Parallel()
+
+				testNamespace := testutils.GenerateTestNamespaceName(t, fmt.Sprintf("%s%s", testNamespacePrefix, tcName), 4)
+				logger := log.WithName(tcName).WithValues("etcdName", defaultEtcdName, "namespace", testNamespace)
+
+				initializeTestCase(g, testEnv, logger, testNamespace, defaultEtcdName)
+
+				logger.Info("running tests")
+				etcdBuilder := testutils.EtcdBuilderWithoutDefaults(defaultEtcdName, testNamespace).
+					WithReplicas(tc.replicas).
+					WithDefaultBackup().
+					WithStorageProvider(provider, fmt.Sprintf("%s/%s", testNamespace, defaultEtcdName))
+
+				if tc.tlsEnabled {
+					etcdBuilder = etcdBuilder.WithClientTLS().
+						WithPeerTLS().
+						WithBackupRestoreTLS()
+				}
+				etcd := etcdBuilder.Build()
+
+				logger.Info("creating Etcd")
+				testEnv.CreateAndCheckEtcd(g, etcd, timeoutEtcdReconciliation)
+				logger.Info("successfully created Etcd")
+
+				if tc.replicas != 0 {
+					logger.Info("hibernating Etcd")
+					replicas := etcd.Spec.Replicas
+					testEnv.HibernateAndCheckEtcd(g, etcd, timeoutEtcdReconciliation)
+					logger.Info("successfully hibernated Etcd")
+
+					logger.Info("unhibernating Etcd")
+					etcd, err := testEnv.GetEtcd(etcd.Name, etcd.Namespace)
+					g.Expect(err).ToNot(HaveOccurred())
+					testEnv.UnhibernateAndCheckEtcd(g, etcd, replicas, timeoutEtcdReconciliation)
+					logger.Info("successfully unhibernated Etcd")
+				}
+
+				logger.Info("deleting Etcd")
+				etcd, err := testEnv.GetEtcd(etcd.Name, etcd.Namespace)
+				g.Expect(err).ToNot(HaveOccurred())
+				testEnv.DeleteAndCheckEtcd(g, logger, etcd, timeoutEtcdReconciliation)
+				logger.Info("successfully deleted Etcd")
+
+				logger.Info("finished running tests")
+
+				if !retainTestArtifacts {
+					cleanupTestArtifactsIfNecessary(testEnv, logger, g, testNamespace, etcd, timeoutEtcdReconciliation)
+				}
```

Comment:
This could be extracted into a function called `cleanupTestArtificacts` and called with a defer statement at the beginning of the test. The check could be done in just one place within the extracted function.